### PR TITLE
Feature modelselection forward

### PR DESCRIPTION
**Pull request by dilpath**
Forward/backward model selection with AIC. Example usage can be found in the tests `test_pipeline_forward` and `test_pipeline_backward`.

TODO:
- [x] BIC
- [x] Custom initial models
- [ ] Use previous MLE as a startpoint for subsequently tested models

### CODE REVIEW

File: pypesto/model_selection/__init__.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+"""
+Model Selection
+========
```
Comment (yannikschaelte) [inline pypesto/model_selection/__init__.py:3]:
```suggestion
Model Selection
===============
```

File: pypesto/model_selection/__init__.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+
+"""
+
+#from .model_selection import ModelSelectionProblem, ModelSelection
```
Comment (yannikschaelte) [inline pypesto/model_selection/__init__.py:8]:
```suggestion
```

File: pypesto/model_selection/__init__.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+from .model_selection import (
+    ModelSelectionProblem,
+    ModelSelector,
+    unpack_file,
+    #ModelSelectorMethod,
```
Comment (yannikschaelte) [inline pypesto/model_selection/__init__.py:13]:
```suggestion
```

File: pypesto/model_selection/constants.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+#YAML_FILENAME_COLUMN = "SBML"
+#MODEL_NAME_COLUMN = "modelId"
```
Comment (yannikschaelte) [inline pypesto/model_selection/constants.py:2]:
```suggestion
# YAML_FILENAME_COLUMN = "SBML"
# MODEL_NAME_COLUMN = "modelId"
```

Comment (yannikschaelte) [reply pypesto/model_selection/constants.py:2]: ore remove ...

File: pypesto/model_selection/model_selection.py [hunk 5]
```diff
@@ -0,0 +1,906 @@
...
+import math
+from ..problem import Problem
+from ..optimize import minimize
+from ..result import Result
+from ..petab import PetabImporter
+from ..objective import Objective
+
+#from .row2problem import row2problem
+
+# TODO replace * with constant names
```
Comment (yannikschaelte) [inline pypesto/model_selection/model_selection.py:17]: TODO

File: pypesto/model_selection/model_selection.py [hunk 6]
```diff
@@ -0,0 +1,906 @@
...
+import petab
+import numpy as np
+import math
+from ..problem import Problem
+from ..optimize import minimize
+from ..result import Result
+from ..petab import PetabImporter
+from ..objective import Objective
+
+#from .row2problem import row2problem
```
Comment (yannikschaelte) [inline pypesto/model_selection/model_selection.py:15]: ?

File: pypesto/model_selection/model_selection.py [hunk 7]
```diff
@@ -0,0 +1,906 @@
...
+# TODO replace * with constant names
+#from .constants import MODEL_NAME_COLUMN, YAML_FILENAME_COLUMN
+from .constants import *
+
+from petab.C import NOMINAL_VALUE, ESTIMATE
+
+
+class ModelSelectionProblem:
+    """
+
```
Comment (yannikschaelte) [inline pypesto/model_selection/model_selection.py:26]: TODO add docstring, short description

File: pypesto/model_selection/model_selection.py [hunk 8]
```diff
@@ -0,0 +1,906 @@
...
+            If `False`, the model parameters will not be estimated. Allows
+            users to manually call pypesto.minimize with custom options, then
+            `set_result()`.
+        fixed_estimated:
+            Parameters that can be fixed to different values can be considered
+            estimated, as the "best" fixed parameter will be preferred. Note,
+            the preference is implemented as comparison of different models
+            with `ModelSelectionMethod.compare()`, unlike normal estimation,
+            which occurs within the same model with `pypesto.minimize`.
+        TODO: constraints
```
Comment (yannikschaelte) [inline pypesto/model_selection/model_selection.py:56]: TODO

File: pypesto/model_selection/model_selection.py [hunk 9]
```diff
@@ -0,0 +1,906 @@
...
+        """
+        self.row = row
+        self.petab_problem = petab_problem
+        self.valid = valid
+
+        if x_fixed_estimated is None:
+            x_fixed_estimated = set()
+        else:
+            # TODO remove parameters that are zero
```
Comment (yannikschaelte) [inline pypesto/model_selection/model_selection.py:65]: TODO

File: pypesto/model_selection/model_selection.py [hunk 10]
```diff
@@ -0,0 +1,906 @@
...
+
+    def set_result(self, result: Result):
+        self.minimize_result = result
+        self.optimized_model = self.minimize_result.optimize_result.list[0]
+
+    @property
+    def AIC(self):
+        if '_AIC' not in dir(self):
+            self._AIC = calculate_AIC(self.n_estimated,
+                                      self.optimized_model.fval)
```
Comment (yannikschaelte) [inline pypesto/model_selection/model_selection.py:102]: usually all attributes should be set in the `__init__` method already. Here you can check for `None` instead.

File: pypesto/model_selection/model_selection.py [hunk 11]
```diff
@@ -0,0 +1,906 @@
...
+                        exclusions: List[str] = None
+    ) -> Iterable[Dict[str, Union[str, float]]]:
+        """
+        A generator for the models described by the model specification file.
+
+        Argument
+        --------
+        #index:
+        #    If None, all models from the model specification file are yielded.
+        #    If not None, only the model at line number `index` is returned.
```
Comment (yannikschaelte) [inline pypesto/model_selection/model_selection.py:193]: ?

File: pypesto/model_selection/model_selection.py (hunk 8 above)
Comment (yannikschaelte) [reply pypesto/model_selection/model_selection.py:56]: what would constraints be, e.g.?

File: pypesto/model_selection/model_selection.py [hunk 12]
```diff
@@ -0,0 +1,906 @@
...
+                self.petab_problem.parameter_df.query(f'{ESTIMATE} == 1').index
+            )
+            self.n_estimated = len(self.estimated)
+            # TODO for BIC. Could move to `calculate_BIC()` to save time.
+            self.n_edata = None
+
+            self.minimize_result = None
+
+            if autorun:
+                self.set_result(minimize(self.pypesto_problem))
```
Comment (yannikschaelte) [inline pypesto/model_selection/model_selection.py:92]: allow to pass a `pypesto.Optimizer` and `n_starts` (or, more generally, the same arguments that are usually passed to pypesto.minimize)?

File: pypesto/model_selection/model_selection.py [hunk 13]
```diff
@@ -0,0 +1,906 @@
...
+        self.optimized_model = self.minimize_result.optimize_result.list[0]
+
+    @property
+    def AIC(self):
+        if '_AIC' not in dir(self):
+            self._AIC = calculate_AIC(self.n_estimated,
+                                      self.optimized_model.fval)
+        return self._AIC
+
+    def calculate_BIC(self):
```
Comment (yannikschaelte) [inline pypesto/model_selection/model_selection.py:105]: calculate_BIC should, like calculate_AIC, be a globally available function

File: pypesto/model_selection/model_selection.py [hunk 14]
```diff
@@ -0,0 +1,906 @@
...
+
+    #def calibrate(self):
+    #    # TODO settings?
+    #    # 100 starts, SingleCoreEngine, ScipyOptimizer
+    #    result = minimize(self.pypesto_problem)
+    #    self.set_result(result)
+
+
+class ModelSelector:
+    def __init__(
```
Comment (yannikschaelte) [inline pypesto/model_selection/model_selection.py:122]: docstring, what it is and does

File: pypesto/model_selection/model_selection.py [hunk 15]
```diff
@@ -0,0 +1,906 @@
...
+        #    self.parse_constraints_file[constraints_filename])
+
+        self.selection_history = {}
+
+        # Dictionary of method names as keys, with a dictionary as the values.
+        # In the dictionary, keys will be modelId, criterion value
+        # TODO unused
+        self.results = {}
+
+    def delete_models(self,
```
Comment (yannikschaelte) [inline pypesto/model_selection/model_selection.py:146]: if the constraints are to be checked in here, better call it `delete_if_constraints_fail` or sth like that.

File: pypesto/model_selection/model_selection.py [hunk 16]
```diff
@@ -0,0 +1,906 @@
...
+        elif method == 'backward':
+            selector = ForwardSelector(self.petab_problem,
+                                       self.model_generator,
+                                       criterion,
+                                       self.parameter_IDs,
+                                       self.selection_history,
+                                       reverse=True)
+            result, self.selection_history = selector()
+        else:
+            raise NotImplementedError(f'Model selection algorithm: {method}.')
```
Comment (yannikschaelte) [inline pypesto/model_selection/model_selection.py:260]:
```suggestion
        if method == 'forward':
            reverse = False
        elif method == 'backward':
            reverse = True
        else:
            raise NotImplementedError(f'Model selection algorithm: {method}.')
        selector = ForwardSelector(
            self.petab_problem,
            self.model_generator,
            criterion,
            self.parameter_IDs,
            self.selection_history,
            reverse=reverse)
        result, self.selection_history = selector()
```

File: pypesto/model_selection/model_selection.py [hunk 17]
```diff
@@ -0,0 +1,906 @@
...
+
+        # TODO: Reconsider return value. `result` could be stored in attribute,
+        # then no values need to be returned, and users can request values
+        # manually.
+        return result, self.selection_history
+
+
+class ModelSelectorMethod:
+    # the calling child class should have self.criterion defined
```
Comment (yannikschaelte) [inline pypesto/model_selection/model_selection.py:269]: add docstring what it is and does

File: pypesto/model_selection/model_selection.py [hunk 18]
```diff
@@ -0,0 +1,906 @@
...
+        else:
+            raise NotImplementedError(f'Model selection algorithm: {method}.')
+
+        # TODO: Reconsider return value. `result` could be stored in attribute,
+        # then no values need to be returned, and users can request values
+        # manually.
+        return result, self.selection_history
+
+
+class ModelSelectorMethod:
```
Comment (yannikschaelte) [inline pypesto/model_selection/model_selection.py:268]: This is an abstract base class, right? I.e. not intended to be used on its own. Then write `class ModelSelectorMethod(abc.ABC)` and indicate all abstract methods that subclasses have to overwrite by te decorator `@abc.abstractmethod`.

File: pypesto/model_selection/constants.py (hunk 4 above)
Comment (MerktSimon) [reply pypesto/model_selection/constants.py:2]: should be removed since it is taken care of by lines 4,6,7

**Pull request closed without merging**